src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs

```diff
@@ -0,0 +1,776 @@
+use crate::{
+    ActivateResult, ConfigResult, DbsGuestAddressSpace, Error as VirtioError,
+    Result as VirtioResult, VirtioDevice, VirtioDeviceConfig, VirtioDeviceInfo, TYPE_NET,
+};
+
+const NET_DRIVER_NAME: &str = "vhost-user-net";
+// Epoll token for incoming connection on the Unix Domain Socket listener.
+const LISTENER_SLOT: u32 = 0;
+// Epoll token for monitoring the Unix Domain Socket between the master and
+// the slave.
+const MASTER_SLOT: u32 = 1;
+// Epoll token for control queue
+const CTRL_SLOT: u32 = 2;
+// Control queue count
+const CTRL_QUEUE_NUM: u16 = 64;
+
+/// An implementation of vhost-user-net device
+struct VhostUserNetDevice {
+    /// Fixed value: "vhost-user-net".
+    id: String,
+    device_info: VirtioDeviceInfo,
+    /// Unix domain socket connecting to the vhost-user slave.
+    endpoint: Endpoint,
+    /// Unix domain socket listener to accept incoming connection from the slave.
+    listener: Listener,
+    /// current enabled queues with vhost-user slave
+    curr_queues: u32,
+}
+
+impl VhostUserNetDevice {
+    fn new(
+        master: Master,
+        mut avail_features: u64,
+        listener: Listener,
+        guest_mac: Option<&MacAddr>,
+        queue_sizes: Arc<Vec<u16>>,
+        epoll_mgr: EpollManager,
+    ) -> VirtioResult<Self> {
+        // hard-coding MTU
+        info!(
+            "{}: slave support features 0x{:x}",
+            NET_DRIVER_NAME, avail_features
+        );
+        avail_features |= (1 << VIRTIO_NET_F_MTU) as u64;
+        // All these features depends on availability of control
+        // channel (VIRTIO_NET_F_CTRL_VQ).
+        avail_features &= !(1 << VIRTIO_NET_F_CTRL_VQ
+            | 1 << VIRTIO_NET_F_CTRL_RX
+            | 1 << VIRTIO_NET_F_CTRL_VLAN
+            | 1 << VIRTIO_NET_F_GUEST_ANNOUNCE
+            | 1 << VIRTIO_NET_F_MQ
+            | 1 << VIRTIO_NET_F_CTRL_MAC_ADDR) as u64;
+        // Multi-queue features
+        if queue_sizes.len() > 2 {
+            avail_features |= (1 << VIRTIO_NET_F_MQ | 1 << VIRTIO_NET_F_CTRL_VQ) as u64;
+        }
+        // Network device configuration layout:
+        // https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-2000004
+        // - [u8; 6]: mac address
+        // - u16: status
+        // - u16: max_virtqueue_pairs
+        // - u16: mtu
+        // - u32: speed
+        // - u8: duplex
+        let mut config_space = vec![0u8; 17];
+        if let Some(mac) = guest_mac {
+            // When this feature isn't available, the driver generates a random
+            // MAC address. Otherwise, it should attempt to read the device MAC
+            // address from the config space.
+            avail_features |= 1u64 << VIRTIO_NET_F_MAC;
+            config_space[0..6].copy_from_slice(mac.get_bytes());
+        } else {
+            avail_features &= !(1 << VIRTIO_NET_F_MAC) as u64;
+        }
+        // status: mark link as up
```

Comment:
could the index here (6,7,8,9..16) be a const so that it's clearer which bit it refers to.

Reply:
Fixed.